src/main/java/org/elasticsearch/search/MultiValueMode.java

```diff
@@ -438,39 +441,65 @@ public long get(int docID) {
      *
      * NOTE: Calling the returned instance on docs that are not root docs is illegal
      */
-    // TODO: technically innerDocs need not be BitSet: only needs advance() ?
-    public NumericDocValues select(final SortedNumericDocValues values, final long missingValue, final BitSet rootDocs, final BitSet innerDocs, int maxDoc) {
-        if (rootDocs == null || innerDocs == null) {
+    public NumericDocValues select(final SortedNumericDocValues values, final long missingValue, final BitSet rootDocs, final DocIdSet innerDocSet, int maxDoc) throws IOException {
+        if (rootDocs == null || innerDocSet == null) {
             return select(DocValues.emptySortedNumeric(maxDoc), missingValue);
         }
+        final DocIdSetIterator innerDocs = innerDocSet.iterator();
+        if (innerDocs == null) {
+            return select(DocValues.emptySortedNumeric(maxDoc), missingValue);
+        }
+
         return new NumericDocValues() {
 
+            int lastSeenRootDoc = -1;
+            long lastEmittedValue;
```

Comment:
Should we rather initialize lastSeenRootDoc=0 and lastEmittedValue=missingValue to avoid the `if (rootDoc == 0)` check below?